src/main/scala/io/iohk/ethereum/network/PeersInfoHolderActor.scala

```diff
@@ -0,0 +1,189 @@
+package io.iohk.ethereum.network
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Props}
+import io.iohk.ethereum.db.storage.AppStateStorage
+import io.iohk.ethereum.network.PeerActor.{DisconnectPeer, SendMessage}
+import io.iohk.ethereum.network.PeersInfoHolderActor._
+import io.iohk.ethereum.network.PeerEventBusActor.{PeerSelector, Subscribe, Unsubscribe}
+import io.iohk.ethereum.network.PeerEventBusActor.PeerEvent._
+import io.iohk.ethereum.network.PeerEventBusActor.SubscriptionClassifier._
+import io.iohk.ethereum.network.handshaker.Handshaker.HandshakeResult
+import io.iohk.ethereum.network.p2p.Message
+import io.iohk.ethereum.network.p2p.messages.CommonMessages.{NewBlock, Status}
+import io.iohk.ethereum.network.p2p.messages.PV62.{BlockHeaders, GetBlockHeaders, NewBlockHashes}
+import io.iohk.ethereum.network.p2p.messages.WireProtocol.Disconnect
+
+class PeersInfoHolderActor(peerEventBusActor: ActorRef, appStateStorage: AppStateStorage,
+                           forkResolverOpt: Option[ForkResolver]) extends Actor with ActorLogging {
+
+  val msgCodesWithInfo: Set[Int] = Set(BlockHeaders.code, NewBlock.code, NewBlockHashes.code)
+
+  peerEventBusActor ! Subscribe(PeerHandshaked)
+
+  override def receive: Receive = handlePeersInfoEvents(Map.empty)
+
+  def handlePeersInfoEvents(peersWithInfo: Map[PeerId, PeerWithInfo]): Receive = {
+
+    case MessageToPeer(message, peerId) if peersWithInfo.contains(peerId) =>
+      val PeerWithInfo(peer, oldPeerInfo) = peersWithInfo(peerId)
+      val newPeerInfo = handleSentMessage(message, oldPeerInfo)
+      context become handlePeersInfoEvents(peersWithInfo + (peerId -> PeerWithInfo(peer, newPeerInfo)))
+
+    case MessageFromPeer(message, peerId) if peersWithInfo.contains(peerId) =>
```

Comment:
This case and the other one before are quite the same. Do you think we can, in some way, to share it to reduce some code?